log4j-core/src/main/java/org/apache/logging/log4j/core/async/RingBufferLogEventHandler4.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.core.async;
+
+import com.lmax.disruptor.EventHandler;
+import com.lmax.disruptor.Sequence;
+import org.apache.logging.log4j.status.StatusLogger;
+import org.apache.logging.log4j.util.LoaderUtil;
+
+/**
+ * This event handler gets passed messages from the RingBuffer as they become
+ * available. Processing of these messages is done in a separate thread,
+ * controlled by the {@code Executor} passed to the {@code Disruptor}
+ * constructor.
+ */
+class RingBufferLogEventHandler4 implements EventHandler<RingBufferLogEvent> {
+
+    private static final int NOTIFY_PROGRESS_THRESHOLD = 50;
+    private Sequence sequenceCallback;
+    private int counter;
+    private long threadId = -1;
+
+    /**
+     * Returns the appropriate {@link EventHandler} for the version of LMAX Disruptor used.
+     */
+    public static RingBufferLogEventHandler4 create() {
+        try {
+            return LoaderUtil.newInstanceOf("org.apache.logging.log4j.core.async.RingBufferLogEventHandler");
+        } catch (final ReflectiveOperationException | LinkageError e) {
+            StatusLogger.getLogger().debug("LMAX Disruptor 3.x is missing, trying version 4.x.", e);
+        }
+        return new RingBufferLogEventHandler4();
```

Comment:
My comment above for `AsyncLoggerConfigDisruptor.createEventHandler()` applies here too.